File: cpp/twelve-days/twelve_days.cpp
#include "twelve_days.h"
#include <vector>

namespace twelve_days
{
    static const std::vector<std::string> DAYS = {
        "first",
        "second",
        "third",
        "fourth",
        "fifth",
        "sixth",
        "seventh",
        "eighth",
        "ninth",
        "tenth",
        "eleventh",
        "twelfth",
    };

    static const std::vector<std::string> GIFTS = {
        "a Partridge in a Pear Tree",
        "two Turtle Doves",
        "three French Hens",
        "four Calling Birds",
        "five Gold Rings",
        "six Geese-a-Laying",
        "seven Swans-a-Swimming",
        "eight Maids-a-Milking",
        "nine Ladies Dancing",
        "ten Lords-a-Leaping",
        "eleven Pipers Piping",
        "twelve Drummers Drumming",
    };

    std::string verse(int verse_number)
    {
        std::string result{};

        result += "On the " + DAYS[verse_number - 1] + " day of Christmas my true love gave to me: ";

        for (int i = verse_number - 1; i >= 0; i--)
        {
            if (i == 0 && verse_number > 1)
            {
                result += "and ";
            }

            result += GIFTS[i];

            if (i > 0)
            {
                result += ", ";
            }
        }

        result += ".\n";

        return result;
    }

    std::string recite(int start_verse, int end_verse)
    {
        std::string result{};

        for (int i = start_verse; i <= end_verse; i++)
        {
            result += verse(i);

            if (i != end_verse)
            {
                result += "\n";
            }
        }

        return result;
    }

} // namespace twelve_days

File: java/twelve-days/src/main/java/TwelveDays.java
import java.util.List;

class TwelveDays {

    private static final List<String> days = List.of(
            "first",
            "second",
            "third",
            "fourth",
            "fifth",
            "sixth",
            "seventh",
            "eighth",
            "ninth",
            "tenth",
            "eleventh",
            "twelfth"
    );

    private static final List<String> gifts = List.of(
            "a Partridge in a Pear Tree",
            "two Turtle Doves",
            "three French Hens",
            "four Calling Birds",
            "five Gold Rings",
            "six Geese-a-Laying",
            "seven Swans-a-Swimming",
            "eight Maids-a-Milking",
            "nine Ladies Dancing",
            "ten Lords-a-Leaping",
            "eleven Pipers Piping",
            "twelve Drummers Drumming"
    );

    String verse(int verseNumber) {
        var sb = new StringBuilder();

        sb.append("On the %s day of Christmas my true love gave to me: ".formatted(days.get(verseNumber - 1)));

        for (var i = verseNumber - 1; i >= 0; i--) {
            if (i == 0 && verseNumber > 1) sb.append("and ");

            sb.append(gifts.get(i));

            if (i > 0) sb.append(", ");
        }

        sb.append('.');
        sb.append('\n');

        return sb.toString();
    }

    String verses(int startVerse, int endVerse) {
        var sb = new StringBuilder();

        for (int i = startVerse; i <= endVerse; i++) {
            sb.append(verse(i));

            if (i != endVerse) sb.append('\n');
        }

        return sb.toString();
    }

    String sing() {
        return verses(1, 12);
    }
}
